cmd/gcp-controller-manager/app/node_annotater.go

```diff
@@ -207,3 +248,40 @@ func parseNodeURL(nodeURL string) (project, zone, instance string, err error) {
 	instance = parts[2]
 	return
 }
+
+// TODO: move this to instance.Labels. This is gross.
+func extractKubeLabels(instance *compute.Instance) (map[string]string, error) {
+	const labelsKey = "kube-labels"
+
+	if instance.Metadata == nil {
+		return nil, fmt.Errorf("instance %q has no labels")
+	}
+
+	var kubeLabels *string
+	for _, item := range instance.Metadata.Items {
+		if item == nil || item.Key != labelsKey {
+			continue
+		}
+		if item.Value == nil {
+			return nil, fmt.Errorf("instance %q had nil %q", instance.SelfLink, labelsKey)
+		}
+		kubeLabels = item.Value
+	}
+	if kubeLabels == nil {
+		return nil, fmt.Errorf("instance %q had no %q", instance.SelfLink, labelsKey)
+	}
+
+	labels := make(map[string]string)
+	for _, kv := range strings.Split(*kubeLabels, ",") {
```

Comment:
This looks pretty fragile. Are label names and values guaranteed to not have commas?

Reply:
Seems like the values can have commas, though this was how we composite labels in various places (e.g. `--node-labels` flag for kubelet and cluster autoscaler).

Reply:
They can't:

https://github.com/kubernetes/kubernetes/blob/4378a999e19aa6ebaa062db5aab84f173d451b12/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go#L88

> a valid label must be an empty string or consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character